### DISCUSSION TITLE
Read Grain tracker object in tensor mechanics

### DISCUSSION THREAD

**Discussion by ritamcj** [idaholab/moose#19616] [Q&A General] [answered]
Hi everyone,
I wanted to ask whether it is possible to read grain tracker inside tensor mechanics. Grain tracker inherits from a file such as Graintrackerinterface, which is present in the phase field module. When I tried `_grain_tracker(getUserObject<GrainTrackerInterface>("grain_tracker"))` in a file in tensor mechanics, I got "fatal error : no such file", which is expected. I also tried `_grain_tracker(getUserObject<GrainTracker>("grain_tracker"))`, as shown in OutputEulerAngles, but this also led to the same error. I wonder whether this is possible by reading as a string in the input file. Kindly suggest strategies or relevant examples for **reading a user object inside a different module**. 
Warm regards,
Ritam

**Comment by GiudGiud** [comment, marked as answer]
Hello

`fatal error : no such file` isn't from `_grain_tracker(getUserObject<GrainTrackerInterface>("grain_tracker"))`. It s probably because a header name was misspelled. 

Could you please paste the entire message?
Also, `_grain_tracker(getUserObject<GrainTracker>("grain_tracker"))` is the correct call. GrainTrackerInterface is an interface. Interfaces are inherited when you need their functionalities, not stored as class member attributes. 

For using objects from a different module, you first need to make sure both modules are being compiled together. This is usually done by creating an application, and adding both modules in the Makefile.

Guillaume

**Reply by ritamcj** [reply]
Hi Guillaume,
Thank you for your reply. I got the following error when I added `#include "GrainTrackerInterface.h"` to `ComputeElasticitytensorCP.C`. I am pasting the error message here:

```
In file included from /home/ritam/projects_new/moose/modules/tensor_mechanics/build/unity_src/materials_crystal_plasticity_Unity.C:2:
/home/ritam/projects_new/moose/modules/tensor_mechanics/src/materials/crystal_plasticity/ComputeElasticityTensorCP.C:12:10: fatal error: GrainTrackerInterface.h: No such file or directory
   12 | #include "GrainTrackerInterface.h"
      |          ^~~~~~~~~~~~~~~~~~~~~~~~~
compilation terminated.
/home/ritam/projects_new/moose/framework/build.mk:122: recipe for target '/home/ritam/projects_new/moose/modules/tensor_mechanics/build/unity_src/materials_crystal_plasticity_Unity.x86_64-conda-linux-gnu.opt.lo' failed
make: *** [/home/ritam/projects_new/moose/modules/tensor_mechanics/build/unity_src/materials_crystal_plasticity_Unity.x86_64-conda-linux-gnu.opt.lo] Error 1
```
 I got the same corresponding error when I used `#include "GrainTracker.h"` along with `_grain_tracker(getUserObject<GrainTracker>("grain_tracker"))`. Could you elaborate on "compiling both modules together"? I am not aware of this. Can I change something in `registerMooseObject("TensorMechanicsApp", ComputeElasticityTensorCP)`? I am currently using the Combined module for running my input file.

Ritam

**Reply by GiudGiud** [reply]
Hello

Both those files exist. The compiler is not finding them at compile time though.
What is your setup? Did you create your own app or are you modifying MOOSE directly?
Are you compiling the combined module directly?

registering objects is a different thing. That s for bridging to the input file. 

Guillaume

**Reply by ritamcj** [reply]
Hi,
I didn't create my own app. I am compiling the combined module directly. 
Ritam

**Reply by GiudGiud** [reply]
Creating your own app is the clean way to do this. 
Now you need to make tensor_mechanics depend on phase_field. I am trying to see if we can get that to compile, since phase_field already depends on tensor_mechanics. 
I ll report back

**Reply by ritamcj** [reply]
Thank you very much. I'll look into how to create apps.

**Reply by GiudGiud** [reply]
Ok can confirm you cant easily make tensor_mechanics depend on phase_field since phase_field depends on tensor_mechanics. Getting some errors at link time. 
The hacky way would be to move the headers you need (GrainTracker.h) in the tensor_mechanics include folder. That way it ll be found by both tensor mechanics and phase field. 
We would rather you create an app and add the phase field dependency in your Makefile.

**Reply by ritamcj** [reply]
Thanks for letting me know. I will try create a separate app.

**Reply by ritamcj** [reply]
Hi Guillaume,
I created a separate app called "DRX" as per https://mooseframework.inl.gov/getting_started/new_users.html#create-an-app. As I already have an input file that uses classes from tensor mechanics and phase field, I think I would have to move all .C and .h files (**from only these two modules**) to DRX and register all classes with the new app. Kindly let me know whether this is correct. Do I also need to move folders with dependent files (such as "framework") to DRX?
Ritam

**Reply by GiudGiud** [reply]
Hello

no you dont need to do that. 
Just make sure MOOSE_DIR is set and in the Makefile for DRX, make sure to put yes for phase_field. Maybe tensor_mechanics as well but phase_field should pull it in already. 

Guillaume

**Reply by ritamcj** [reply]
Hi Guillaume,
It did work. When I hit "make" in the new app, it compiled with Grain tracker inside ComputeElasticitytensorCP. I now realise that I should've always compiled the combined module instead of compiling tensor mechanics and phase field separately. Many thanks for your guidance. I feel enlightened.
Ritam
